csharp/ql/lib/semmle/code/csharp/dataflow/ExternalFlow.qll

```diff
@@ -343,58 +345,123 @@ module CsvValidation {
-      kind = row.splitAt(";", 7) and
-      not kind = ["code", "sql", "xss", "remote", "html"] and
-      not kind.matches("encryption-%") and
-      msg = "Invalid kind \"" + kind + "\" in sink model."
     )
-    or
-    exists(string row, string kind | sourceModel(row) |
-      kind = row.splitAt(";", 7) and
-      not kind = ["local", "file"] and
-      msg = "Invalid kind \"" + kind + "\" in source model."
+  }
+
+  private string getInvalidModelSignature() {
+    exists(
+      string pred, string namespace, string type, string name, string signature, string ext,
+      string provenance
+    |
+      sourceModel(namespace, type, _, name, signature, ext, _, _, provenance) and pred = "source"
+      or
+      sinkModel(namespace, type, _, name, signature, ext, _, _, provenance) and pred = "sink"
+      or
+      summaryModel(namespace, type, _, name, signature, ext, _, _, _, provenance) and
+      pred = "summary"
+      or
+      negativeSummaryModel(namespace, type, name, signature, provenance) and
+      ext = "" and
+      pred = "negative summary"
+    |
+      not namespace.regexpMatch("[a-zA-Z0-9_\\.]+") and
+      result = "Dubious namespace \"" + namespace + "\" in " + pred + " model."
+      or
+      not type.regexpMatch("[a-zA-Z0-9_<>,\\+]+") and
+      result = "Dubious type \"" + type + "\" in " + pred + " model."
+      or
+      not name.regexpMatch("[a-zA-Z0-9_<>,]*") and
+      result = "Dubious member name \"" + name + "\" in " + pred + " model."
+      or
+      not signature.regexpMatch("|\\([a-zA-Z0-9_<>\\.\\+\\*,\\[\\]]*\\)") and
+      result = "Dubious signature \"" + signature + "\" in " + pred + " model."
+      or
+      not ext.regexpMatch("|Attribute") and
+      result = "Unrecognized extra API graph element \"" + ext + "\" in " + pred + " model."
+      or
+      not provenance = ["manual", "generated"] and
+      result = "Unrecognized provenance description \"" + provenance + "\" in " + pred + " model."
     )
   }
+
+  /** Holds if some row in a CSV-based flow model appears to contain typos. */
+  query predicate invalidModelRow(string msg) {
+    msg =
+      [
+        getInvalidModelSignature(), getInvalidModelInput(), getInvalidModelOutput(),
+        getInvalidModelSubtype(), getInvalidModelColumnCount(), getInvalidModelKind()
+      ]
+  }
 }
 
 private predicate elementSpec(
   string namespace, string type, boolean subtypes, string name, string signature, string ext
 ) {
-  sourceModel(namespace, type, subtypes, name, signature, ext, _, _, _) or
-  sinkModel(namespace, type, subtypes, name, signature, ext, _, _, _) or
+  sourceModel(namespace, type, subtypes, name, signature, ext, _, _, _)
+  or
+  sinkModel(namespace, type, subtypes, name, signature, ext, _, _, _)
+  or
   summaryModel(namespace, type, subtypes, name, signature, ext, _, _, _, _)
+  or
+  negativeSummaryModel(namespace, type, name, signature, _) and ext = "" and subtypes = false
```

Comment:
This makes anything depending on the other kinds of models also depend on negative summaries. This might be fine for now, but if they grow too large that we don't want to scan them for regular queries then we'd likely need to split this predicate here (and do the implementation sharing bits with parameterised modules instead). But it's probably fine to leave that as potential future work.

Reply:
The idea is that the modules containing the negative summaries are included near the queries that relies on them. Are you worried that if a query that imports the negative summaries are included in query package containing many other queries that these other queries will be slower?

Reply:
Currently inclusion is based on file module activation (aka transitive import), and the negative models aren't activated for regular queries, so they'll be fine. But all of the positive models and their dependencies will be recomputed once negative models are included, so any queries referring to negative models need to do extra work. Again, this is probably fine for now.
I originally didn't notice that the negative models were excluded from the computation by way of not being activated, since the picture in my mind was a more explicit split at the predicate level. And once we move the MaD rows to external files then we need to refactor this, since at that point the exclusion-by-no-import won't work.

Reply:
Great! Thank you very much for the explanation! 😃